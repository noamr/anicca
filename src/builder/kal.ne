main -> (statements)                           {% ([s]) => JSON.parse(JSON.stringify(s)) %}

@include "./formula.ne"
constValue -> 
    primitive {% ([$primitive]) => ({$primitive}) %}
    | qvar {% ([t]) => ({$ref: t.value, $token: extractToken(t)}) %}

newlines -> %eol {% NOOP %}
    | %eol newlines {% NOOP %}

maybeNewlines -> newlines {% NOOP %} 
    | null {% NOOP %}

beforeChildren -> _ newlines %indent {% NOOP %}
afterChildren -> maybeNewlines %dedent {% NOOP %}
IndentChildren[X] -> beforeChildren $X afterChildren {% ([,[children]]) => children %}

DeclareKeyword[Keyword] -> $Keyword __ {% extractToken %}
Conjunction[Keyword] -> __ $Keyword __ {% NOOP %}

Statement[Keyword, Children] ->
    DeclareKeyword[$Keyword] %varname IndentChildren[$Children] 
        {% ([token, name, [children]]) => ({...token, name: name.value, children}) %}

@{% function Statement(type, childrenKey) 
    { return ([{$token, name, children}]) => ({$token, type, name, [childrenKey]: children}) } %}

betweenChildren -> _ maybeNewlines _
ChildrenOfType[X, XX] ->
    $X {% id %}
| $X betweenChildren $XX {% ([[one], , [additional]]) => ([one, ...additional]) %}


statements -> ChildrenOfType[anyStatement, statements] {% id %}

anyStatement -> 
    slotStatement {% id %}
    | mainStatement  {% id %}     
    | busStatement {% id %}                  
    | constStatement {% id %}                     
    | letStatement {% id %}                
    | macroStatement {% id %}
    | importStatement {% id %}                     
    | viewStatement {% id %}    
    | controllerStatement {% id %}                  

viewStatement -> Statement["view", viewRules] {% Statement('View', 'rules') %}
busStatement -> DeclareKeyword["bus"] %varname _ newlines {% ([token, name]) => ({"type": "Bus", "name": name.value, $token: extractToken(token)}) %}
slotStatement -> Statement["slot", formula] {% Statement('Slot', 'formula') %}
importStatement -> DeclareKeyword["import"] _ stringLiteral Conjunction["as"] %varname WS {% ([token,, path,, name]) => ({type: "Import", path: eval(path), name: name.value, $token: extractToken(token)}) %}

macroStatement -> DeclareKeyword["macro"] %varname _ "(" _ macroArgs _ ")" IndentChildren[formula]
    {% ([,t,,,,args,,,formula]) => ({$macro: {args, formula}, $token: extractToken(t)}) %}

macroArgs ->
    null {% () => [] %}
    | %varname {% ([{value}]) => [value] %}
    | %varname _ "," _ macroArgs {% ([one,,,, additionalArgs]) => ([one.value, ...additionalArgs]) %}

selectorValue ->
    %varname {%id%}
    | stringLiteral {%id%}
    | number {%id%}

selectorAttribComp ->
    "=" {% id %}
    | "~=" {% id %}
    | "*=" {% id %}
    | "$=" {% id %}
    | "^=" {% id %}

selectorAttribs ->
    _ "[" _ %varname _ selectorAttribComp _ selectorValue _ "]" {% (a) => a.join('') %}
    | _ "[" _ %varname _ "]" {% (a) => a.join('') %}

selectorKey ->
    "*" {%id%}
    | %varname {% id %}
    | %selector {% id%}

selector ->
    selectorKey {% id %}
    | selectorKey selectorAttribs {% ([key, attribs]) => ({...key, value: key.value + attribs}) %}



viewRules -> ChildrenOfType[viewRule, viewRules] {% id %}
viewRule -> selector IndentChildren[viewDeclarations]
    {% ([selector, declarations]) => ({type: 'ViewRule', declarations, selector: selector.value, $token: extractToken(selector)})  %}

viewDeclarations -> ChildrenOfType[viewDeclaration, viewDeclarations] {% id %}

viewDeclaration ->
    viewBindDeclaration {% id %}
    | viewEventDeclaration {% id %}

viewBindDeclaration ->
    DeclareKeyword["bind"] viewBindTarget to formula
        {% ([token, target, , src]) => ({...token, type: "Bind", target, src}) %}

viewBindTarget ->
    "html" {% ([t]) => ({type: 'html', $token: extractToken(t)}) %}

viewEventDeclaration ->
    DeclareKeyword["on"] %varname IndentChildren[viewEventActions]
        {% ([token, eventType, actions]) => ({type: "DOMEvent", $token: extractToken(token), eventType: eventType.value, actions}) %}

viewEventActions -> ChildrenOfType[viewEventAction, viewEventActions] {% id %}

viewEventAction ->
    viewEventDispatchAction {% id %}
    | preventDefault {% id %}

to -> Conjunction["to"] {% NOOP %}

controllerStatement -> Statement["controller", statechart] {% Statement("Controller", "statechart") %}

statechart -> branchState {% ([root]) => ({root}) %}
maybeStateName ->
    %varname {% ([name]) => ({name: name.value, $token: extractToken(name)}) %}
    | null {% NOOP %}

exclusiveStateHeader ->
    DeclareKeyword["state"] %varname {% ([t, name]) => ({$token: extractToken(t), name: name.value}) %}
    | %varname {% ([name]) => ({$token: extractToken(name), name: name.value}) %}

exclusiveState -> exclusiveStateHeader IndentChildren[stateChildren] {% ([header, children]) => ({type: 'State', ...header, children}) %}
branchState -> exclusiveState {% id %}
state -> branchState {% id %}
stateChild -> 
    state {% id %}
    | transition {% id %}

stateChildren -> ChildrenOfType[stateChild, stateChildren] {% id %}

transition ->
    conditionalTransition {% id %}

conditionalTransition ->
    transitionHeaders IndentChildren[transitionActions] {% ([header, actions]) => 
        ({type: 'Transition', ...header.reduce((a, o) => Object.assign(a, o), {}), actions}) %}

transitionHeaders -> ChildrenOfType[transitionHeader, transitionHeaders] {% id %}
transitionHeader ->
    transitionCondition {% id %}
    | transitionTrigger {% id %}
    | transitionTarget {% id %}

transitionTarget ->
    DeclareKeyword["goto"] qvar {% ([$token, event]) => ({event: {$ref: event.value}, $token}) %}

transitionCondition ->
    DeclareKeyword["when"] formula {% ([$token, condition]) => ({$token, condition}) %}

transitionTrigger ->
    DeclareKeyword["on"] qvar {% ([$token, event]) => ({event: {$ref: event.value}, $token}) %}

transitionActions -> ChildrenOfType[transitionAction, transitionActions] {% id %}
transitionAction ->
    incrementAction {% id %}
    | assignAction {% id %}
    | decrementAction {% id %}
    | multAction {% id %}
    | divAction {% id %}

Operand[O] => _ $O _ {% ([, [op]]) => op %}

@{%
    const relAction = op => ([target,,,, src]) => 
        ({type: "Assign", target: {$ref: target.value}, src: {op, $token: extractToken(target), args: [
            {$ref: target.value, $token: extractToken(target)}, src
        ]}})
%}

incrementAction ->
    qvar _ "+=" _ formula {% relAction('add') %}

decrementAction ->
    qvar _ "-=" _ formula {% relAction('add') %}

multAction ->
    qvar _ "*=" _ formula {% relAction('div') %}

divAction ->
    qvar _ "/=" _ formula {% relAction('div') %}

assignAction ->
    qvar _ "=" _ formula {% ([target,, op,, src]) => 
        ({type: "Assign", target: {$ref: target.value}, src}) %}

viewEventDispatchAction ->
    DeclareKeyword["dispatch"] qvar to qvar
        {% ([token, event, , bus]) => ({$token: extractToken(token), event: event.value, bus: bus.value, type: 'Dispatch'}) %}

preventDefault ->
    "prevent" " " "default" {% t => ({type: "PreventDefault", $token: extractToken(t)}) %}

formula -> rawFormula {% id %}

VarDeclaration[Type] ->
    DeclareKeyword[$Type] %varname Conjunction["="] constValue WS {% ([type, name,, value]) => 
        ({name: name.value, value, $token: extractToken(type)})%} 

constStatement -> VarDeclaration["const"] {% ([v]) => ({...v, type: 'Const'}) %}
letStatement -> VarDeclaration["let"] {% ([v]) => ({...v, type: 'Let'}) %}

mainStatement -> 
    "main" _ mainChildren WS
        {% ([token,, declarations]) => ({$token: extractToken(token), type: 'Main', declarations}) %}
mainChildren -> ChildrenOfType[useDeclaration, mainChildren] {% id %}

useDeclaration ->
    DeclareKeyword["use"] qvar {% ([token, name]) => ({type: "Use", ref: name.value, ...token}) %}

