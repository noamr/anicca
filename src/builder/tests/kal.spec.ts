/// </ reference "jest" />
import nearley from 'nearley'
const kalGrammar = require('../kal.ne')
import {
    Formula,
    FunctionFormula,
    PrimitiveFormula,
    ReferenceFormula,
    Bundle
} from '../types'

const parse = (str: string): Bundle => new nearley.Parser(kalGrammar).feed(str).finish()[0]
const removeTokens = (a: any): any =>
    a && Array.isArray(a) ? a.map(removeTokens) : Object.keys(a)
    .map(k => k === '$token' ? {} : {
        [k]: a[k]
    })
    .reduce((a, o) => Object.assign(a, o), {})

const parseRaw = (str: string) => removeTokens(parse(str))
describe('kal', () => {
    describe('formulas', () => {
        it('slot', () => {
            expect(parse(`
        slot s
            a + 1
        `)).toMatchSnapshot()
        })
        it('slot 2', () => {
            expect(parse(`
        slot Abc1
            (a + 1) ** (13 / 7 & X) |> sin()
        `)).toMatchSnapshot()
        })
        it('slot 3', () => {
            expect(parse(`
        slot pipe
            2 |> sin()
        `)).toMatchSnapshot()
        })
    })

    describe('main', () => {
        it('main', () => {
            expect(parse(`
            main
                use hello
                use view

    `)).toMatchSnapshot()
        })
    })

    describe('bus', () => {
        it('bus', () => {
            expect(parse(`
            bus myBus
    `)).toMatchSnapshot()

        })
    })
    describe('vars', () => {
        it('let', () => {
            expect(parse(`
            let abc = 1
    `)).toMatchSnapshot()

        })
        it('const', () => {
            expect(parse(`
            const hello = 1
            let abc = '123'
            let ahf = 'blabla'
    `)).toMatchSnapshot()

        })
        it('with refs', () => {
            expect(parse(`
            const hello = abc
            let abc = abc:a$0
    `)).toMatchSnapshot()

        })
    })

    describe('macro', () => {
        it('no args', () => {
            expect(parse(`
            macro abc()
                a + 1
    `)).toMatchSnapshot()
        })
        it('some args', () => {
            expect(parse(`
            macro abc(a)
                a + 1
    `)).toMatchSnapshot()
        })
        it('several args', () => {
            expect(parse(`
            macro abc(a, b)
                a + 13 * b
    `)).toMatchSnapshot()
        })
    })

    describe('files', () => {
        it('import', () => {
            expect(parse(`
            import 'hello.kal' as hello
            `)).toMatchSnapshot()
        })
    })

    describe('view', () => {
        it('all kinda stuff', () => {
            expect(parse(`
            view myView
                #someId
                    bind html to someText
                    on click
                        dispatch someEvent to someController
                
            `)).toMatchSnapshot()
        })
        it('strange selectors', () => {
            expect(parse(`
            view myView
                div
                    on click
                        prevent default
                div[ab=123]
                    on click
                        dispatch someEvent to someABC
                        prevent default
                *[yo]
                    bind html to 123
                
            `)).toMatchSnapshot()
        })
    })

    describe('controller', () => {
        it('long statechart', () => {
            expect(parse(`
            controller myController
                state someState
                    anotherState
                        on click
                            a+= 1
                    on someEvent
                        abc = 3
            `)).toMatchSnapshot()
        })
        it('actions', () => {
            expect(parse(`
            controller myController
                state someState
                    on event
                        a -= 1
                        dda:abcvb /= 4
                        foo:bar *= 2
            `)).toMatchSnapshot()
        })
        it('on when', () => {
            expect(parse(`
            controller myController
                state someState
                    on event
                        a -= 1
                    on event when condition
                        a += 1
                    on event when !bla
                        a += 1
            `)).toMatchSnapshot()
        })
        it('when without on', () => {
            expect(parse(`
            controller myController
                state someState
                    when condition
                        a -= 123
                    when func(something)
                        a += 1
                    on event when b |> max(a)
                        a = b
            `)).toMatchSnapshot()
        })
        it('goto', () => {
            expect(parse(`
            controller myController
                state someState
                    goto somewhere
                    when a <= 1
                        x:a = 0
                    goto otherState when something
                        a -= 123
                    goto otherState when a == 1
                        a -= 123
            `)).toMatchSnapshot()
        })
    })
    // when someCondition goto module.someState
    // when someCondition goto someOtherEvent2
    //     a = b
    // on event when condition goto state
    //     a *= sin(3)


})