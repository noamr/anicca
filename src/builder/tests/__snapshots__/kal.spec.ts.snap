// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kal bus bus 1`] = `
Array [
  Array [
    Object {
      "$token": Object {
        "col": 13,
        "line": 2,
      },
      "name": "myBus",
      "type": "Bus",
    },
  ],
]
`;

exports[`kal controller actions 1`] = `
Array [
  Array [
    Object {
      "name": "myController",
      "statechart": Object {
        "root": Object {
          "$token": Object {
            "col": 17,
            "line": 3,
          },
          "children": Array [
            Object {
              "$token": Object {
                "col": 21,
                "line": 4,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 5,
                    },
                    "args": Array [
                      Object {
                        "$ref": "a",
                        "$token": Object {
                          "col": 25,
                          "line": 5,
                        },
                      },
                      Object {
                        "$primitive": 1,
                      },
                    ],
                    "op": "add",
                  },
                  "target": Object {
                    "$ref": "a",
                  },
                  "type": "Assign",
                },
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 6,
                    },
                    "args": Array [
                      Object {
                        "$ref": "dda.abcvb",
                        "$token": Object {
                          "col": 25,
                          "line": 6,
                        },
                      },
                      Object {
                        "$primitive": 4,
                      },
                    ],
                    "op": "div",
                  },
                  "target": Object {
                    "$ref": "dda.abcvb",
                  },
                  "type": "Assign",
                },
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 7,
                    },
                    "args": Array [
                      Object {
                        "$ref": "foo.bar",
                        "$token": Object {
                          "col": 25,
                          "line": 7,
                        },
                      },
                      Object {
                        "$primitive": 2,
                      },
                    ],
                    "op": "div",
                  },
                  "target": Object {
                    "$ref": "foo.bar",
                  },
                  "type": "Assign",
                },
              ],
              "event": Object {
                "$ref": "event",
              },
              "type": "Transition",
            },
          ],
          "name": "someState",
          "type": "State",
        },
      },
      "type": "Controller",
    },
  ],
]
`;

exports[`kal controller goto 1`] = `
Array [
  Array [
    Object {
      "name": "myController",
      "statechart": Object {
        "root": Object {
          "$token": Object {
            "col": 17,
            "line": 3,
          },
          "children": Array [
            Object {
              "$token": Object {
                "col": 21,
                "line": 5,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$primitive": 0,
                  },
                  "target": Object {
                    "$ref": "x.a",
                  },
                  "type": "Assign",
                },
              ],
              "condition": Object {
                "$token": Object {
                  "col": 28,
                  "line": 5,
                },
                "args": Array [
                  Object {
                    "$ref": "a",
                    "$token": Object {
                      "col": 26,
                      "line": 5,
                    },
                  },
                  Object {
                    "$primitive": 1,
                  },
                ],
                "op": "lte",
              },
              "event": Object {
                "$ref": "somewhere",
              },
              "type": "Transition",
            },
            Object {
              "$token": Object {
                "col": 37,
                "line": 7,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 8,
                    },
                    "args": Array [
                      Object {
                        "$ref": "a",
                        "$token": Object {
                          "col": 25,
                          "line": 8,
                        },
                      },
                      Object {
                        "$primitive": 123,
                      },
                    ],
                    "op": "add",
                  },
                  "target": Object {
                    "$ref": "a",
                  },
                  "type": "Assign",
                },
              ],
              "condition": Object {
                "$ref": "something",
                "$token": Object {
                  "col": 42,
                  "line": 7,
                },
              },
              "event": Object {
                "$ref": "otherState",
              },
              "type": "Transition",
            },
            Object {
              "$token": Object {
                "col": 37,
                "line": 9,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 10,
                    },
                    "args": Array [
                      Object {
                        "$ref": "a",
                        "$token": Object {
                          "col": 25,
                          "line": 10,
                        },
                      },
                      Object {
                        "$primitive": 123,
                      },
                    ],
                    "op": "add",
                  },
                  "target": Object {
                    "$ref": "a",
                  },
                  "type": "Assign",
                },
              ],
              "condition": Object {
                "$token": Object {
                  "col": 44,
                  "line": 9,
                },
                "args": Array [
                  Object {
                    "$ref": "a",
                    "$token": Object {
                      "col": 42,
                      "line": 9,
                    },
                  },
                  Object {
                    "$primitive": 1,
                  },
                ],
                "op": "eq",
              },
              "event": Object {
                "$ref": "otherState",
              },
              "type": "Transition",
            },
          ],
          "name": "someState",
          "type": "State",
        },
      },
      "type": "Controller",
    },
  ],
]
`;

exports[`kal controller long statechart 1`] = `
Array [
  Array [
    Object {
      "name": "myController",
      "statechart": Object {
        "root": Object {
          "$token": Object {
            "col": 17,
            "line": 3,
          },
          "children": Array [
            Object {
              "$token": Object {
                "col": 21,
                "line": 4,
              },
              "children": Array [
                Object {
                  "$token": Object {
                    "col": 25,
                    "line": 5,
                  },
                  "actions": Array [
                    Object {
                      "src": Object {
                        "$token": Object {
                          "col": 29,
                          "line": 6,
                        },
                        "args": Array [
                          Object {
                            "$ref": "a",
                            "$token": Object {
                              "col": 29,
                              "line": 6,
                            },
                          },
                          Object {
                            "$primitive": 1,
                          },
                        ],
                        "op": "add",
                      },
                      "target": Object {
                        "$ref": "a",
                      },
                      "type": "Assign",
                    },
                  ],
                  "event": Object {
                    "$ref": "click",
                  },
                  "type": "Transition",
                },
              ],
              "name": "anotherState",
              "type": "State",
            },
            Object {
              "$token": Object {
                "col": 21,
                "line": 7,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$primitive": 3,
                  },
                  "target": Object {
                    "$ref": "abc",
                  },
                  "type": "Assign",
                },
              ],
              "event": Object {
                "$ref": "someEvent",
              },
              "type": "Transition",
            },
          ],
          "name": "someState",
          "type": "State",
        },
      },
      "type": "Controller",
    },
  ],
]
`;

exports[`kal controller on when 1`] = `
Array [
  Array [
    Object {
      "name": "myController",
      "statechart": Object {
        "root": Object {
          "$token": Object {
            "col": 17,
            "line": 3,
          },
          "children": Array [
            Object {
              "$token": Object {
                "col": 21,
                "line": 4,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 5,
                    },
                    "args": Array [
                      Object {
                        "$ref": "a",
                        "$token": Object {
                          "col": 25,
                          "line": 5,
                        },
                      },
                      Object {
                        "$primitive": 1,
                      },
                    ],
                    "op": "add",
                  },
                  "target": Object {
                    "$ref": "a",
                  },
                  "type": "Assign",
                },
              ],
              "event": Object {
                "$ref": "event",
              },
              "type": "Transition",
            },
            Object {
              "$token": Object {
                "col": 30,
                "line": 6,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 7,
                    },
                    "args": Array [
                      Object {
                        "$ref": "a",
                        "$token": Object {
                          "col": 25,
                          "line": 7,
                        },
                      },
                      Object {
                        "$primitive": 1,
                      },
                    ],
                    "op": "add",
                  },
                  "target": Object {
                    "$ref": "a",
                  },
                  "type": "Assign",
                },
              ],
              "condition": Object {
                "$ref": "condition",
                "$token": Object {
                  "col": 35,
                  "line": 6,
                },
              },
              "event": Object {
                "$ref": "event",
              },
              "type": "Transition",
            },
            Object {
              "$token": Object {
                "col": 30,
                "line": 8,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 9,
                    },
                    "args": Array [
                      Object {
                        "$ref": "a",
                        "$token": Object {
                          "col": 25,
                          "line": 9,
                        },
                      },
                      Object {
                        "$primitive": 1,
                      },
                    ],
                    "op": "add",
                  },
                  "target": Object {
                    "$ref": "a",
                  },
                  "type": "Assign",
                },
              ],
              "condition": Object {
                "$token": Object {
                  "col": 35,
                  "line": 8,
                },
                "args": Array [
                  Object {
                    "$ref": "bla",
                    "$token": Object {
                      "col": 36,
                      "line": 8,
                    },
                  },
                ],
                "op": "not",
              },
              "event": Object {
                "$ref": "event",
              },
              "type": "Transition",
            },
          ],
          "name": "someState",
          "type": "State",
        },
      },
      "type": "Controller",
    },
  ],
]
`;

exports[`kal controller when without on 1`] = `
Array [
  Array [
    Object {
      "name": "myController",
      "statechart": Object {
        "root": Object {
          "$token": Object {
            "col": 17,
            "line": 3,
          },
          "children": Array [
            Object {
              "$token": Object {
                "col": 21,
                "line": 4,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 5,
                    },
                    "args": Array [
                      Object {
                        "$ref": "a",
                        "$token": Object {
                          "col": 25,
                          "line": 5,
                        },
                      },
                      Object {
                        "$primitive": 123,
                      },
                    ],
                    "op": "add",
                  },
                  "target": Object {
                    "$ref": "a",
                  },
                  "type": "Assign",
                },
              ],
              "condition": Object {
                "$ref": "condition",
                "$token": Object {
                  "col": 26,
                  "line": 4,
                },
              },
              "type": "Transition",
            },
            Object {
              "$token": Object {
                "col": 21,
                "line": 6,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$token": Object {
                      "col": 25,
                      "line": 7,
                    },
                    "args": Array [
                      Object {
                        "$ref": "a",
                        "$token": Object {
                          "col": 25,
                          "line": 7,
                        },
                      },
                      Object {
                        "$primitive": 1,
                      },
                    ],
                    "op": "add",
                  },
                  "target": Object {
                    "$ref": "a",
                  },
                  "type": "Assign",
                },
              ],
              "condition": Object {
                "$token": Object {
                  "col": 26,
                  "line": 6,
                },
                "args": Array [
                  Object {
                    "$ref": "something",
                    "$token": Object {
                      "col": 31,
                      "line": 6,
                    },
                  },
                ],
                "op": "func",
              },
              "type": "Transition",
            },
            Object {
              "$token": Object {
                "col": 30,
                "line": 8,
              },
              "actions": Array [
                Object {
                  "src": Object {
                    "$ref": "b",
                    "$token": Object {
                      "col": 29,
                      "line": 9,
                    },
                  },
                  "target": Object {
                    "$ref": "a",
                  },
                  "type": "Assign",
                },
              ],
              "condition": Object {
                "$token": Object {
                  "col": 37,
                  "line": 8,
                },
                "args": Array [
                  Object {
                    "$ref": "b",
                    "$token": Object {
                      "col": 35,
                      "line": 8,
                    },
                  },
                  Object {
                    "$ref": "a",
                    "$token": Object {
                      "col": 44,
                      "line": 8,
                    },
                  },
                ],
                "op": "max",
              },
              "event": Object {
                "$ref": "event",
              },
              "type": "Transition",
            },
          ],
          "name": "someState",
          "type": "State",
        },
      },
      "type": "Controller",
    },
  ],
]
`;

exports[`kal files import 1`] = `
Array [
  Array [
    Object {
      "$token": Object {
        "col": 13,
        "line": 2,
      },
      "name": "hello",
      "path": "hello.kal",
      "type": "Import",
    },
  ],
]
`;

exports[`kal formulas slot 1`] = `
Array [
  Array [
    Object {
      "formula": Object {
        "$token": Object {
          "col": 15,
          "line": 3,
        },
        "args": Array [
          Object {
            "$ref": "a",
            "$token": Object {
              "col": 13,
              "line": 3,
            },
          },
          Object {
            "$primitive": 1,
          },
        ],
        "op": "add",
      },
      "name": "s",
      "type": "Slot",
    },
  ],
]
`;

exports[`kal formulas slot 2 1`] = `
Array [
  Array [
    Object {
      "formula": Object {
        "$token": Object {
          "col": 37,
          "line": 3,
        },
        "args": Array [
          Object {
            "$token": Object {
              "col": 21,
              "line": 3,
            },
            "args": Array [
              Object {
                "$token": Object {
                  "col": 16,
                  "line": 3,
                },
                "args": Array [
                  Object {
                    "$ref": "a",
                    "$token": Object {
                      "col": 14,
                      "line": 3,
                    },
                  },
                  Object {
                    "$primitive": 1,
                  },
                ],
                "op": "add",
              },
              Object {
                "$token": Object {
                  "col": 32,
                  "line": 3,
                },
                "args": Array [
                  Object {
                    "$token": Object {
                      "col": 28,
                      "line": 3,
                    },
                    "args": Array [
                      Object {
                        "$primitive": 13,
                      },
                      Object {
                        "$primitive": 7,
                      },
                    ],
                    "op": "div",
                  },
                  Object {
                    "$ref": "X",
                    "$token": Object {
                      "col": 34,
                      "line": 3,
                    },
                  },
                ],
                "op": "bwand",
              },
            ],
            "op": "pow",
          },
        ],
        "op": "sin",
      },
      "name": "Abc1",
      "type": "Slot",
    },
  ],
]
`;

exports[`kal formulas slot 3 1`] = `
Array [
  Array [
    Object {
      "formula": Object {
        "$token": Object {
          "col": 15,
          "line": 3,
        },
        "args": Array [
          Object {
            "$primitive": 2,
          },
        ],
        "op": "sin",
      },
      "name": "pipe",
      "type": "Slot",
    },
  ],
]
`;

exports[`kal macro no args 1`] = `
Array [
  Array [
    Object {
      "$macro": Object {
        "args": Array [],
        "formula": Object {
          "$token": Object {
            "col": 19,
            "line": 3,
          },
          "args": Array [
            Object {
              "$ref": "a",
              "$token": Object {
                "col": 17,
                "line": 3,
              },
            },
            Object {
              "$primitive": 1,
            },
          ],
          "op": "add",
        },
      },
      "$token": Object {
        "col": 19,
        "line": 2,
      },
    },
  ],
]
`;

exports[`kal macro several args 1`] = `
Array [
  Array [
    Object {
      "$macro": Object {
        "args": Array [
          "a",
          "b",
        ],
        "formula": Object {
          "$token": Object {
            "col": 19,
            "line": 3,
          },
          "args": Array [
            Object {
              "$ref": "a",
              "$token": Object {
                "col": 17,
                "line": 3,
              },
            },
            Object {
              "$token": Object {
                "col": 24,
                "line": 3,
              },
              "args": Array [
                Object {
                  "$primitive": 13,
                },
                Object {
                  "$ref": "b",
                  "$token": Object {
                    "col": 26,
                    "line": 3,
                  },
                },
              ],
              "op": "mult",
            },
          ],
          "op": "add",
        },
      },
      "$token": Object {
        "col": 19,
        "line": 2,
      },
    },
  ],
]
`;

exports[`kal macro some args 1`] = `
Array [
  Array [
    Object {
      "$macro": Object {
        "args": Array [
          "a",
        ],
        "formula": Object {
          "$token": Object {
            "col": 19,
            "line": 3,
          },
          "args": Array [
            Object {
              "$ref": "a",
              "$token": Object {
                "col": 17,
                "line": 3,
              },
            },
            Object {
              "$primitive": 1,
            },
          ],
          "op": "add",
        },
      },
      "$token": Object {
        "col": 19,
        "line": 2,
      },
    },
  ],
]
`;

exports[`kal main main 1`] = `
Array [
  Array [
    Object {
      "$token": Object {
        "col": 13,
        "line": 2,
      },
      "declarations": Array [
        Object {
          "col": 17,
          "line": 3,
          "ref": "hello",
          "type": "Use",
        },
        Object {
          "col": 17,
          "line": 4,
          "ref": "view",
          "type": "Use",
        },
      ],
      "type": "Main",
    },
  ],
]
`;

exports[`kal vars const 1`] = `
Array [
  Array [
    Object {
      "$token": Object {
        "col": 13,
        "line": 2,
      },
      "name": "hello",
      "type": "Const",
      "value": Object {
        "$primitive": 1,
      },
    },
    Object {
      "$token": Object {
        "col": 13,
        "line": 3,
      },
      "name": "abc",
      "type": "Let",
      "value": Object {
        "$primitive": "123",
      },
    },
    Object {
      "$token": Object {
        "col": 13,
        "line": 4,
      },
      "name": "ahf",
      "type": "Let",
      "value": Object {
        "$primitive": "blabla",
      },
    },
  ],
]
`;

exports[`kal vars let 1`] = `
Array [
  Array [
    Object {
      "$token": Object {
        "col": 13,
        "line": 2,
      },
      "name": "abc",
      "type": "Let",
      "value": Object {
        "$primitive": 1,
      },
    },
  ],
]
`;

exports[`kal vars with refs 1`] = `
Array [
  Array [
    Object {
      "$token": Object {
        "col": 13,
        "line": 2,
      },
      "name": "hello",
      "type": "Const",
      "value": Object {
        "$ref": "abc",
        "$token": Object {
          "col": 27,
          "line": 2,
        },
      },
    },
    Object {
      "$token": Object {
        "col": 13,
        "line": 3,
      },
      "name": "abc",
      "type": "Let",
      "value": Object {
        "$ref": "abc.a$0",
        "$token": Object {
          "col": 23,
          "line": 3,
        },
      },
    },
  ],
]
`;

exports[`kal view all kinda stuff 1`] = `
Array [
  Array [
    Object {
      "name": "myView",
      "rules": Array [
        Object {
          "$token": Object {
            "col": 17,
            "line": 3,
          },
          "declarations": Array [
            Object {
              "col": 21,
              "line": 4,
              "src": Object {
                "$ref": "someText",
                "$token": Object {
                  "col": 34,
                  "line": 4,
                },
              },
              "target": Object {
                "$token": Object {
                  "col": 26,
                  "line": 4,
                },
                "type": "html",
              },
              "type": "Bind",
            },
            Object {
              "$token": Object {
                "col": 21,
                "line": 5,
              },
              "actions": Array [
                Object {
                  "$token": Object {
                    "col": 25,
                    "line": 6,
                  },
                  "bus": "someController",
                  "event": "someEvent",
                  "type": "Dispatch",
                },
              ],
              "eventType": "click",
              "type": "DOMEvent",
            },
          ],
          "selector": "#someId",
          "type": "ViewRule",
        },
      ],
      "type": "View",
    },
  ],
]
`;

exports[`kal view strange selectors 1`] = `
Array [
  Array [
    Object {
      "name": "myView",
      "rules": Array [
        Object {
          "$token": Object {
            "col": 17,
            "line": 3,
          },
          "declarations": Array [
            Object {
              "$token": Object {
                "col": 21,
                "line": 4,
              },
              "actions": Array [
                Object {
                  "$token": Object {
                    "col": 25,
                    "line": 5,
                  },
                  "type": "PreventDefault",
                },
              ],
              "eventType": "click",
              "type": "DOMEvent",
            },
          ],
          "selector": "div",
          "type": "ViewRule",
        },
        Object {
          "$token": Object {
            "col": 17,
            "line": 6,
          },
          "declarations": Array [
            Object {
              "$token": Object {
                "col": 21,
                "line": 7,
              },
              "actions": Array [
                Object {
                  "$token": Object {
                    "col": 25,
                    "line": 8,
                  },
                  "bus": "someABC",
                  "event": "someEvent",
                  "type": "Dispatch",
                },
                Object {
                  "$token": Object {
                    "col": 25,
                    "line": 9,
                  },
                  "type": "PreventDefault",
                },
              ],
              "eventType": "click",
              "type": "DOMEvent",
            },
          ],
          "selector": "div[ab=123]",
          "type": "ViewRule",
        },
        Object {
          "$token": Object {
            "col": 17,
            "line": 10,
          },
          "declarations": Array [
            Object {
              "col": 21,
              "line": 11,
              "src": Object {
                "$primitive": 123,
              },
              "target": Object {
                "$token": Object {
                  "col": 26,
                  "line": 11,
                },
                "type": "html",
              },
              "type": "Bind",
            },
          ],
          "selector": "*[yo]",
          "type": "ViewRule",
        },
      ],
      "type": "View",
    },
  ],
]
`;
